site/_posts/2020-08-20-building-profile-portal-with-flask-oauth-apis.md

```diff
@@ -0,0 +1,815 @@
+      {% endif %}
+      {% if key == 'minprice' %}
+        Minimum home price: <input type='{{fields['registration.data.'+key].control}}' {% if fields['registration.data.'+key].required %}required{% endif %} name='{{key}}' value='{{registration_data['minprice']}}' />
+      {% endif %}
+      </p>
+    {% endfor %}
+  {% endif %}
+  <input type="submit" value="Update"></input>
+</form>
+<!-- ... -->
+```
+
+This builds a form from the same `registration_data` array. But it does something else which you might find intriguing. Let's take a deeper look at the maximum price form field, which has some additional newlines added below:
+
+```html
+<!-- ... -->
+Maximum home price: 
+  <input 
+  type='{{fields['registration.data.'+key].control}}' 
+  {% if fields['registration.data.'+key].required %}required{% endif %} 
+  name='{{key}}' 
+  value='{{registration_data['maxprice']}}' />
+<!-- ... -->
+```
+
+How are the `type` and `required` attributes determined? What is this `fields` value? 
+
+The source of this mysterious `fields` value are the form field definitions. If the field definition changes, the form field generated is modified as well.
+
+Let's look at the code, where `registration_data` and `fields` variables are set up. Here's the entire `oauth.py` file, but below we'll examine the changes:
+
+```python
+from requests_oauthlib import OAuth2Session
+from flask import Flask, request, redirect, session, url_for, render_template
+from flask.json import jsonify
+from fusionauth.fusionauth_client import FusionAuthClient
+
+import json
+import os
+
+app = Flask(__name__)
+app.secret_key = os.urandom(24)
+app.config.from_object('settings.Config')
+
+@app.route('/', methods=["GET"])
+def homepage():
+  user=None
+  registration_data=None
+  fields = {}
+  if session.get('user') != None:
+    user = session['user']
+    fusionauth_api_client = FusionAuthClient(app.config['API_KEY'], app.config['FA_URL'])
+    user_id = user['sub']
+    application_id = user['applicationId']
+    client_response = fusionauth_api_client.retrieve_registration(user_id, application_id)
+    if client_response.was_successful():
+      #print(client_response.success_response)
+      registration_data = client_response.success_response['registration'].get('data')
+      fields = get_fields(fusionauth_api_client)
+    else:
+      print(client_response.error_response)
+  return render_template('index.html', user=user, registration_data=registration_data, fields=fields)
+
+@app.route("/update", methods=["POST"])
+def update():
+  user=None
+  error=None
+  fields=[]
+  fusionauth_api_client = FusionAuthClient(app.config['API_KEY'], app.config['FA_URL'])
+  if session.get('user') != None:
+    user = session['user']
+    user_id = user['sub']
+    application_id = user['applicationId']
+
+    client_response = fusionauth_api_client.retrieve_registration(user_id, application_id)
+    if client_response.was_successful():
+      #print(client_response.success_response)
```

Comment:
Do we want to leave these print lines in here commented out? There are a couple in this code snippet.

Reply:
Will remove.